SPO/README.md

```diff
@@ -0,0 +1,103 @@
+
+## Set up Application Insights 
+1. Follow [SPFx Extensions instruction](https://docs.microsoft.com/en-us/sharepoint/dev/spfx/extensions/get-started/build-a-hello-world-extension) to create a SharePoint Framework Extension.
+2. Install Application Insights via [NPM setup](https://github.com/microsoft/ApplicationInsights-JS#getting-started). 
+3. Under your Extension solution src folder, update ApplicationCustomizer.ts file:
+    a. [Set up Application Insights](https://github.com/microsoft/ApplicationInsights-JS#npm-setup-ignore-if-using-snippet-setup)
+    ```js
+    import { ApplicationInsights } from '@microsoft/applicationinsights-web'
+
+    const appInsights = new ApplicationInsights({ config: {
+    instrumentationKey: 'YOUR_INSTRUMENTATION_KEY_GOES_HERE'
+    /* ...Other Configuration Options... */
+    } });
+    appInsights.loadAppInsights();
+    appInsights.trackPageView(); // Manually call trackPageView to establish the current user/session/pageview
+    ```
+
+    b. Expose all configuration options you need in the ApplicationCustomizerProperties interface.
+    ```js
+    export interface IAppInsightsApplicationCustomizerProperties {
+        /* ...Other Configuration Options... */
+        instrumentationKey: string;
+    }
+    ```
+    c. Update appInsights instance config to map with exposed properties.
+    ```js
+    import { ApplicationInsights } from '@microsoft/applicationinsights-web'
+
+    const appInsights = new ApplicationInsights({ config: {
+    instrumentationKey: this.properties.instrumentationKey
+    /* ...Other Configuration Options... */
+    } });
+    appInsights.loadAppInsights();
+    appInsights.trackPageView(); // Manually call trackPageView to establish the current user/session/pageview
+    ```
+## Set up deployment configuration
+### Deploy Extension solution Tenant wide (All sites share same iKey)
+1. Under your Extension solution config folder, set the skipFeatureDeployment attribute to true in the package-solution.json file. And add "ClientSideInstance.xml" file under "features"/"assets"/"elementManifests".
+<p><img src="./img/image3.png"/></p>
+
+2. Export properties you defined from ClientSideInstance.xml file.
+    ```js
+    <?xml version="1.0" encoding="utf-8"?>
+    <Elements xmlns="http://schemas.microsoft.com/sharepoint/">
+        <ClientSideComponentInstance
+            Title="AppinsightsExtension"
+            Location="ClientSideExtension.ApplicationCustomizer"
+            ComponentId="96262ed0-fe5d-4786-abc2-9f6b9cb3fefc"
+            Properties="{&quot;instrumentationKey&quot;:&quot;YOUR_INSTRUMENTATION_KEY_GOES_HERE&quot;}">
+        </ClientSideComponentInstance>
+    </Elements>
+    ```
+    
+3. Run following command to create .sppkg file. Check [Host extension from CDN](https://docs.microsoft.com/en-us/sharepoint/dev/spfx/extensions/get-started/hosting-extension-from-office365-cdn) for reference.
```

Comment:
This is a lot of steps. What do you think about creating a SKPPG file for OOB enablement, then reference full doc for anyone who wants custom? I don't want to create more to maintain, but it also seems like some customers may give up.

Reply:
Yeah they have many moving parts. SPO have 3 files for the same properties... (where you link the iKey and other configs to your solution):

- If you build locally, edit `serve.json` -> `properties` (I didn't add this part in README)
- If you deploy to single site, edit `element.xml` -> `ClientSideComponentProperties` 
- If you deploy tenant wide, edit `ClientSideInstance.xml` -> `Properties` 

And they are under different folders. The only way I can think of is to upload sppkg file for them to download, but in that way user will lose the flexibility to config appInsights and add their custom data, which will still raise supportability I guess.

Reply:
Thanks Xiao. I actually like your meta explanation above. Should we work this in to docs?

Reply:
Yeah that sounds reasonable. I'll add it and will tag you :) Will need your help to address language.